File: benchmarks/livermore_canonical/k18_hydro_2d.c
/* K18 -- 2-D explicit hydrodynamics fragment (canonical, 1-based indexing)
   Canonical body: lines 438-466 of the kernels listing.
   T=0.0037, S=0.0041 hardcoded in canonical pre-loop.
   JN set to n-1 (=100) so j+1 access stays within 101-wide fast dim.
   ZM gets +10 offset to keep divisor bounded away from 0.
   Column-major: arr(j,k) at flat (k-1)*101 + j. */
#include <stdio.h>
#include <time.h>
#include "signel.h"

#define N     101
#define NREPS 2570000

#define ZA(j,k) za[((k)-1)*101 + (j)]
#define ZB(j,k) zb[((k)-1)*101 + (j)]
#define ZP(j,k) zp[((k)-1)*101 + (j)]
#define ZQ(j,k) zq[((k)-1)*101 + (j)]
#define ZR(j,k) zr[((k)-1)*101 + (j)]
#define ZM(j,k) zm[((k)-1)*101 + (j)]
#define ZU(j,k) zu[((k)-1)*101 + (j)]
#define ZV(j,k) zv[((k)-1)*101 + (j)]
#define ZZ(j,k) zz[((k)-1)*101 + (j)]

int main(void) {
    static double za[101*7 + 1], zb[101*7 + 1], zp[101*7 + 1], zq[101*7 + 1];
    static double zr[101*7 + 1], zm[101*7 + 1], zu[101*7 + 1], zv[101*7 + 1];
    static double zz[101*7 + 1];

    double fuzz, buzz, fizz;
    fuzz = 1.2345e-3; buzz = 1.0 + fuzz; fizz = 1.1 * fuzz;
    for (long j = 1; j <= 7; j++)
        for (long i = 1; i <= 101; i++) {
            buzz = (1.0 - fuzz) * buzz + fuzz; fuzz = -fuzz;
            ZP(i,j) = (buzz - fizz) * 0.1;
        }
    fuzz = 1.2345e-3; buzz = 1.0 + fuzz; fizz = 1.1 * fuzz;
    for (long j = 1; j <= 7; j++)
        for (long i = 1; i <= 101; i++) {
            buzz = (1.0 - fuzz) * buzz + fuzz; fuzz = -fuzz;
            ZQ(i,j) = (buzz - fizz) * 0.1;
        }
    fuzz = 1.2345e-3; buzz = 1.0 + fuzz; fizz = 1.1 * fuzz;
    for (long j = 1; j <= 7; j++)
        for (long i = 1; i <= 101; i++) {
            buzz = (1.0 - fuzz) * buzz + fuzz; fuzz = -fuzz;
            ZR(i,j) = (buzz - fizz) * 0.1;
        }
    fuzz = 1.2345e-3; buzz = 1.0 + fuzz; fizz = 1.1 * fuzz;
    for (long j = 1; j <= 7; j++)
        for (long i = 1; i <= 101; i++) {
            buzz = (1.0 - fuzz) * buzz + fuzz; fuzz = -fuzz;
            ZM(i,j) = (buzz - fizz) * 0.1 + 10.0;
        }
    fuzz = 1.2345e-3; buzz = 1.0 + fuzz; fizz = 1.1 * fuzz;
    for (long j = 1; j <= 7; j++)
        for (long i = 1; i <= 101; i++) {
            buzz = (1.0 - fuzz) * buzz + fuzz; fuzz = -fuzz;
            ZZ(i,j) = (buzz - fizz) * 0.1;
        }

    for (long j = 1; j <= 7; j++)
        for (long i = 1; i <= 101; i++) {
            ZA(i,j) = 0.0;
            ZB(i,j) = 0.0;
            ZU(i,j) = 0.0;
            ZV(i,j) = 0.0;
        }

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    double t, s;
    long kn, jn;
    for (long rep = 1; rep <= NREPS; rep++) {
        t = 0.0037; s = 0.0041;
        kn = 6; jn = N - 1;
        for (long k = 2; k <= kn; k++)
            for (long j = 2; j <= jn; j++) {
                ZA(j,k) = (ZP(j-1,k+1) + ZQ(j-1,k+1)
                           - ZP(j-1,k) - ZQ(j-1,k))
                         * (ZR(j,k) + ZR(j-1,k))
                         / (ZM(j-1,k) + ZM(j-1,k+1));
                ZB(j,k) = (ZP(j-1,k) + ZQ(j-1,k)
                           - ZP(j,k) - ZQ(j,k))
                         * (ZR(j,k) + ZR(j,k-1))
                         / (ZM(j,k) + ZM(j-1,k));
            }
        for (long k = 2; k <= kn; k++)
            for (long j = 2; j <= jn; j++) {
                ZU(j,k) = ZU(j,k) + s*(ZA(j,k)*(ZZ(j,k)-ZZ(j+1,k))
                                      -ZA(j-1,k)*(ZZ(j,k)-ZZ(j-1,k))
                                      -ZB(j,k)*(ZZ(j,k)-ZZ(j,k-1))
                                      +ZB(j,k+1)*(ZZ(j,k)-ZZ(j,k+1)));
                ZV(j,k) = ZV(j,k) + s*(ZA(j,k)*(ZR(j,k)-ZR(j+1,k))
                                      -ZA(j-1,k)*(ZR(j,k)-ZR(j-1,k))
                                      -ZB(j,k)*(ZR(j,k)-ZR(j,k-1))
                                      +ZB(j,k+1)*(ZR(j,k)-ZR(j,k+1)));
            }
        for (long k = 2; k <= kn; k++)
            for (long j = 2; j <= jn; j++) {
                ZR(j,k) = ZR(j,k) + t*ZU(j,k);
                ZZ(j,k) = ZZ(j,k) + t*ZV(j,k);
            }
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) * 1e-9;
    printf("elapsed: %.6f s\n", elapsed);
    printf("zu(51,4) = %.15e\n", ZU(51,4));
    return 0;
}

File: benchmarks/livermore_canonical/k18_hydro_2d.f
C     Kernel 18 -- 2-D explicit hydrodynamics fragment (canonical)
C     From "The Livermore Fortran Kernels" (UCRL-53745), F.H. McMahon, 1986
C     Canonical body: lines 438-466 of the kernels listing.
C     T=0.0037, S=0.0041 hardcoded in canonical pre-loop.
C     JN set to n-1 (=100) so j+1 access stays within 101-wide fast dim.
C     ZM gets +10 offset to keep divisor bounded away from 0.
C
      PROGRAM K18
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      INTEGER REP
      DIMENSION ZA(101,7), ZB(101,7), ZP(101,7), ZQ(101,7)
      DIMENSION ZR(101,7), ZM(101,7), ZU(101,7), ZV(101,7), ZZ(101,7)
C
C     ZP via SIGNEL
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 12 J = 1, 7
         DO 11 I = 1, 101
            BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
            FUZZ = -FUZZ
            ZP(I,J) = (BUZZ - FIZZ) * 0.1D0
   11    CONTINUE
   12 CONTINUE
C
C     ZQ via SIGNEL
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 14 J = 1, 7
         DO 13 I = 1, 101
            BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
            FUZZ = -FUZZ
            ZQ(I,J) = (BUZZ - FIZZ) * 0.1D0
   13    CONTINUE
   14 CONTINUE
C
C     ZR via SIGNEL
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 16 J = 1, 7
         DO 15 I = 1, 101
            BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
            FUZZ = -FUZZ
            ZR(I,J) = (BUZZ - FIZZ) * 0.1D0
   15    CONTINUE
   16 CONTINUE
C
C     ZM via SIGNEL + 10 (avoid divisor near zero)
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 18 J = 1, 7
         DO 17 I = 1, 101
            BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
            FUZZ = -FUZZ
            ZM(I,J) = (BUZZ - FIZZ) * 0.1D0 + 10.0D0
   17    CONTINUE
   18 CONTINUE
C
C     ZZ via SIGNEL
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 22 J = 1, 7
         DO 21 I = 1, 101
            BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
            FUZZ = -FUZZ
            ZZ(I,J) = (BUZZ - FIZZ) * 0.1D0
   21    CONTINUE
   22 CONTINUE
C
C     Zero ZA, ZB, ZU, ZV
C
      DO 26 J = 1, 7
         DO 25 I = 1, 101
            ZA(I,J) = 0.0D0
            ZB(I,J) = 0.0D0
            ZU(I,J) = 0.0D0
            ZV(I,J) = 0.0D0
   25    CONTINUE
   26 CONTINUE
C
      N = 101
C
C     Kernel 18 body (canonical lines 438-466)
C
      DO 500 REP = 1, 3840000
         T  = 0.003700D0
         S  = 0.004100D0
         KN = 6
         JN = N - 1
         DO 70 K = 2, KN
            DO 170 J = 2, JN
               ZA(J,K) = (ZP(J-1,K+1)+ZQ(J-1,K+1)
     &                    -ZP(J-1,K)-ZQ(J-1,K))
     &                  * (ZR(J,K)+ZR(J-1,K))
     &                  / (ZM(J-1,K)+ZM(J-1,K+1))
               ZB(J,K) = (ZP(J-1,K)+ZQ(J-1,K)
     &                    -ZP(J,K)-ZQ(J,K))
     &                  * (ZR(J,K)+ZR(J,K-1))
     &                  / (ZM(J,K)+ZM(J-1,K))
  170       CONTINUE
   70    CONTINUE
C
         DO 72 K = 2, KN
            DO 172 J = 2, JN
               ZU(J,K) = ZU(J,K) + S*(ZA(J,K)*(ZZ(J,K)-ZZ(J+1,K))
     &                              -ZA(J-1,K)*(ZZ(J,K)-ZZ(J-1,K))
     &                              -ZB(J,K)*(ZZ(J,K)-ZZ(J,K-1))
     &                              +ZB(J,K+1)*(ZZ(J,K)-ZZ(J,K+1)))
               ZV(J,K) = ZV(J,K) + S*(ZA(J,K)*(ZR(J,K)-ZR(J+1,K))
     &                              -ZA(J-1,K)*(ZR(J,K)-ZR(J-1,K))
     &                              -ZB(J,K)*(ZR(J,K)-ZR(J,K-1))
     &                              +ZB(J,K+1)*(ZR(J,K)-ZR(J,K+1)))
  172       CONTINUE
   72    CONTINUE
C
         DO 75 K = 2, KN
            DO 175 J = 2, JN
               ZR(J,K) = ZR(J,K) + T*ZU(J,K)
               ZZ(J,K) = ZZ(J,K) + T*ZV(J,K)
  175       CONTINUE
   75    CONTINUE
  500 CONTINUE
C
      WRITE(*,900) ZU(51,4)
  900 FORMAT('zu(51,4) = ', E25.15)
      END
